### ISSUE TITLE
track_means buffers for PROCAP and ATAC look malformed in lifted over weights

### ISSUE THREAD

**Issue by Al-Murphy** [genomicsxai/alphagenome-pytorch#42]
Hey — heads up on gtca/alphagenome_pytorch::model_all_folds.safetensors (snapshot b01c0ffa…3b9920 on the Hub). The track_means buffers for PROCAP and ATAC look malformed; the other 5 track heads look fine.

Reproducer below (10 s on cpu, uses only alphagenome-pytorch + huggingface_hub, no other wrappers).

Symptoms:

* PROCAP row 0 (should be human, 128 distinct nonzero means) has only 12 nonzero entries, and those 12 are the same 6 values repeated twice — then 116 zeros. Row 1 (should be mouse, ≤32 real means) is entirely 1.0, i.e. fully populated by np.pad(..., constant_values=1.0) from get_track_means_for_head. That's only consistent with metadata.get(OutputType.PROCAP) returning a 12-row dataframe for human and an empty/all-NaN one for mouse at conversion time.
* ATAC row 0 has 167/256 nonzero; row 1 has 18/228 nonzero. Same flavour, less catastrophic.
* Downstream effect: predictions_scaling multiplies the model output by track_means * resolution, so 116/128 human PROCAP tracks unscale to identically zero → zero variance → per-track Pearson collapses to ~0. On the AlphaGenome TEST split I get pooled PROCAP Pearson = 0.06 with full unscale, vs DNASE/ATAC/CAGE/RNA-seq all 0.47–0.63 on the same eval. Bypassing the * track_means step recovers PROCAP, which confirms the model weights themselves are fine — only the means buffer is bad.

Could you check whether metadata.load(Organism.HOMO_SAPIENS).get(OutputType.PROCAP)['nonzero_mean'] actually has 128 finite values at the source, or whether 116 of them are NaN? If the source is fine, the conversion that produced the published safetensors needs a re-run; if the source is NaN-laden, that's the upstream fix. Let me know if you need more info!!

Code:

```
import torch
from huggingface_hub import hf_hub_download
from alphagenome_pytorch import AlphaGenome

weights_path = hf_hub_download(
    repo_id="gtca/alphagenome_pytorch",
    filename="model_all_folds.safetensors",
)
model = AlphaGenome.from_pretrained(weights_path, device="cpu")

# convert_weights.py:91 stacks [human, mouse], so row 0 should be human.
expected = {
    "atac": (256, 228), "dnase": (384, 120), "procap": (128, 32),
    "cage": (640, 226), "rna_seq": (768, 336),
    "chip_tf": (1664, 544), "chip_histone": (1152, 460),
}
print(f"{'head':14s} row0_nonzero   row1_nonzero   row1_all_ones")
print("-" * 60)
for name, (h, m) in expected.items():
    tm = model.heads[name].track_means          # (2, num_tracks)
    nz0 = (tm[0] != 0).sum().item()
    nz1 = (tm[1] != 0).sum().item()
    print(f"{name:14s}   {nz0:>5}/{h:<5}    {nz1:>5}/{m:<5}    "
          f"{bool((tm[1]==1.0).all().item())}")

# PROCAP zoom-in: row 0 has 6 unique values duplicated, then zeros.
tm = model.heads["procap"].track_means
print("\nPROCAP row0[:12] :", tm[0][:12].tolist())
print("PROCAP row0[12:24]:", tm[0][12:24].tolist())
print("PROCAP row1 unique:", torch.unique(tm[1]).tolist())
```

**Comment by gtca** [comment]
Thanks, @Al-Murphy!

From a quick look at it, it seems that the described track means for PRO-cap and ATAC actually match upstream:

```py
import numpy as np, jax.numpy as jnp
from alphagenome.models import dna_output
from alphagenome_research.model import dna_model
from alphagenome_research.model.metadata import metadata as md

# Verbatim _get_track_means body from
# alphagenome_research/model/heads.py:473-488
def jax_track_means(meta, num_tracks):
    if meta is None or meta.get('nonzero_mean') is None:
        return jnp.ones((num_tracks,))
    v = meta['nonzero_mean'].values.reshape(-1)
    return jnp.pad(v, ((0, num_tracks - v.shape[0]),), constant_values=1.0)

md_h = md.load(dna_model.Organism.HOMO_SAPIENS)
md_m = md.load(dna_model.Organism.MUS_MUSCULUS)
df_h = md_h.get(dna_output.OutputType.PROCAP)
df_m = md_m.get(dna_output.OutputType.PROCAP)

print("human PROCAP metadata:")
print(df_h[['name', 'strand', 'nonzero_mean']].iloc[[*range(12), *range(124, 128)]])

tm_h = np.asarray(jax_track_means(df_h, 128))
tm_m = np.asarray(jax_track_means(df_m, 128))
print(f"\nJAX human track_means: finite {np.isfinite(tm_h).sum()}, "
      f"NaN {np.isnan(tm_h).sum()}, ==1.0 {(tm_h == 1.0).sum()}")
print(f"JAX mouse track_means: finite {np.isfinite(tm_m).sum()}, "
      f"NaN {np.isnan(tm_m).sum()}, ==1.0 {(tm_m == 1.0).sum()}")
```

which gives

```
human PROCAP metadata:
            name strand  nonzero_mean
0    ENCSR072YCM      +     13.308302
1    ENCSR182QNJ      +     14.002803
2    ENCSR740IPL      +     15.765458
3    ENCSR797DEF      +     12.281321
4    ENCSR801ECP      +     13.973692
5    ENCSR860TYZ      +      8.229502
6    ENCSR072YCM      -     13.308302
7    ENCSR182QNJ      -     14.002803
8    ENCSR740IPL      -     15.765458
9    ENCSR797DEF      -     12.281321
10   ENCSR801ECP      -     13.973692
11   ENCSR860TYZ      -      8.229502
124      Padding      .           NaN
125      Padding      .           NaN
126      Padding      .           NaN
127      Padding      .           NaN

JAX human track_means: finite 12, NaN 116, ==1.0 0
JAX mouse track_means: finite 128, NaN 0, ==1.0 128
```

Something like this seems to give the same output as your code snippet:

```py
import numpy as np
import jax.numpy as jnp

from alphagenome.models import dna_output
from alphagenome_research.model import dna_model
from alphagenome_research.model.metadata import metadata as metadata_module


# From alphagenome_research/model/heads.py
def _get_track_means(meta, num_tracks):
    if meta is None or meta.get('nonzero_mean') is None:
        return jnp.ones((num_tracks,))
    v = meta['nonzero_mean'].values.reshape(-1)
    # Pad short metadata up to num_tracks with 1.0 (matches the installed
    # version of JAX; the source repo omits this pad but upstream metadata
    # is already num_tracks-shaped so it's a no-op here).
    if v.shape[0] < num_tracks:
        v = jnp.pad(v, ((0, num_tracks - v.shape[0]),), constant_values=1.0)
    return v

expected = {
    "atac":         (256, 228),
    "dnase":        (384, 120),
    "procap":       (128, 32),
    "cage":         (640, 226),
    "rna_seq":      (768, 336),
    "chip_tf":      (1664, 544),
    "chip_histone": (1152, 460),
}

output_type_map = {
    "atac":         dna_output.OutputType.ATAC,
    "dnase":        dna_output.OutputType.DNASE,
    "procap":       dna_output.OutputType.PROCAP,
    "cage":         dna_output.OutputType.CAGE,
    "rna_seq":      dna_output.OutputType.RNA_SEQ,
    "chip_tf":      dna_output.OutputType.CHIP_TF,
    "chip_histone": dna_output.OutputType.CHIP_HISTONE,
}

md_h = metadata_module.load(dna_model.Organism.HOMO_SAPIENS)
md_m = metadata_module.load(dna_model.Organism.MUS_MUSCULUS)

def jax_track_means(head_name, num_tracks):
    ot = output_type_map[head_name]
    tm_h = np.asarray(_get_track_means(md_h.get(ot), num_tracks))
    tm_m = np.asarray(_get_track_means(md_m.get(ot), num_tracks))
    return np.stack([tm_h, tm_m], axis=0)

def finite_nonzero(x):
    finite = np.isfinite(x) & (x != 0)
    nan    = np.isnan(x)
    return int(finite.sum()), int(nan.sum())

print(f"{'head':14s} row0_finite_nz  row1_finite_nz  row0_NaN  row1_NaN  row1_all_ones")
print("-" * 78)
for name, (h, m) in expected.items():
    tm = jax_track_means(name, h)
    nz0, nan0 = finite_nonzero(tm[0])
    nz1, nan1 = finite_nonzero(tm[1])
    all_ones1 = bool(np.all(tm[1] == 1.0))
    print(f"{name:14s}   {nz0:>5}/{h:<5}     {nz1:>5}/{m:<5}     "
          f"{nan0:>5}    {nan1:>5}      {all_ones1}")

tm = jax_track_means("procap", 128)
print("\nPROCAP row0[:12] :", tm[0][:12].tolist())
print("PROCAP row0[12:24]:", tm[0][12:24].tolist())
print("PROCAP row1 unique:", np.unique(tm[1]).tolist())
```

The difference here is that we have `nan_to_num(track_means, nan=0.0)`. If the padding tracks are not masked out, they end up in the metrics calculations, and that is more apparent for the modalities with most tracks being padding, such as PRO-cap.

Would it be better we align and have `NaN`s in the output?

**Comment by Al-Murphy** [comment]
Ahh okay I understand, apologies for that! Up to you with the masking, I guess to me it's fine either way!
